### PR TITLE
Support for Multiple Separate TPU Worker Groups per RayCluster

### PR DESCRIPTION
**Pull request by ryanaoleary**
This PR adds support for multiple worker groups that request TPUs (both single-host and multi-host) in the same Ray Cluster. This PR was tested by manually creating a RayCluster with single-host, multi-host, single-slice, and multi-slice worker groups and verifying that all were injected with the correct `TPU_NAME`, `TPU_WORKER_ID`, `replicaIndex`, and `TPU_WORKER_HOSTNAMES`.

This PR also adds unit tests for `getReplicaIndex` and `getNextWorkerID`, these tests can be run using `go test` from the applications/ray/kuberay-tpu-webhook folder.

### CODE REVIEW

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 1]
```diff
@@ -77,7 +77,7 @@ func containerRequestingTPUs(containers ...corev1.Container) bool {
 	return false
 }
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/main.go:79]: Are `clusterName`, `namespace`, and `groupName` only used for logging? It seems a bit odd to me that these are required parameters for a fairly deterministic function.

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 2]
```diff
@@ -263,22 +265,24 @@ func validateRayCluster(admissionReview *admissionv1.AdmissionReview) (*admissio
...
-			groupName := workerGroupSpec.GroupName
-			podSlice := slice{clusterName, groupName, replicaIndex, numOfHosts}
-			sliceToWorkers[podSlice] = nil
-			sliceToHostnames[podSlice] = ""
-			// generate TPU_WORKER_HOSTNAMES
-			if numOfHosts > 1 {
-				joinedHostNames, err := genDNSHostnames(workerGroupSpec, replicaIndex)
-				if err != nil {
-					klog.Error("Failed to generate DNS Hostnames")
+		if containerRequestingTPUs(workerGroupSpec.Template.Spec.Containers...) {
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/main.go:268]: Log the case where the worker group does not request TPUs?

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 3]
```diff
@@ -320,13 +324,16 @@ func getEnvironmentVariable(varName string, container corev1.Container) string {
...
+func getReplicaIndex(clusterName string, groupName string, namespace string) int {
+	// first pod created in cluster
 	if sliceToWorkers == nil {
 		return 0
 	}
 	nextLowestId := math.MaxInt32
+	numSlices := 0 // tracks # of slices in worker group created so far
 	for slice, workerList := range sliceToWorkers {
-		if slice.clusterName == clusterName {
+		if slice.clusterName == clusterName && slice.groupName == groupName {
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/main.go:335]: Should check for namespace as well?

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 4]
```diff
@@ -320,13 +324,16 @@ func getEnvironmentVariable(varName string, container corev1.Container) string {
 
 // get next lowest-index pod slice to assign a pod to in the RayCluster
 // this will be the first pod slice with # created pods < NumOfHosts
-func getReplicaIndex(clusterName string, namespace string) int {
+func getReplicaIndex(clusterName string, groupName string, namespace string) int {
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/main.go:327]: This logic is a bit complex and is critical to the correctness of the webhook. I would suggest:
1. Clearly document the algorithm used here to pick the next replica index;
2. Create a unit test here that captures all of the corner cases we found so far.

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 5]
```diff
@@ -320,13 +324,16 @@ func getEnvironmentVariable(varName string, container corev1.Container) string {
...
+	// first pod created in cluster
 	if sliceToWorkers == nil {
 		return 0
 	}
 	nextLowestId := math.MaxInt32
+	numSlices := 0 // tracks # of slices in worker group created so far
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/main.go:333]: What does "created" mean in this context? The code here increments numSlices regardless of what happens to it.

Comment (richardsliu) [reply applications/ray/kuberay-tpu-webhook/main.go:333]: Suggest using "replica" instead of "slice" to minimize confusion.

File: applications/ray/kuberay-tpu-webhook/main.go [hunk 6]
```diff
@@ -340,10 +347,11 @@ func getReplicaIndex(clusterName string, namespace string) int {
 			}
 		}
 	}
+	// first pod of new slice in cluster
 	if nextLowestId == math.MaxInt32 {
-		klog.ErrorS(errors.New("Replica Index never set"), "RayCluster", namespace+"/"+clusterName, "Replica Index", nextLowestId)
+		nextLowestId = numSlices
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/main.go:352]: When would we get here? If `sliceToWorkers` is populated when we validate the ray cluster, then shouldn't we always find a match?

File: applications/ray/kuberay-tpu-webhook/main.go (hunk 1 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go:79]: Yeah that's the case for `getNumTPUHostsFromTopology`, the logic behind adding them was so that we could have klog Error and Info statements that would be traceable to the right Group, Namespace, and RayCluster if it failed in this function.

File: applications/ray/kuberay-tpu-webhook/main.go (hunk 6 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go:352]: `sliceToWorkers` isn't populated when we validate the RayCluster, `sliceToHostnames` is. We populate `sliceToWorkers` as `TPU_WORKER_ID`s are assigned in `getNextWorkerID`. We get to this if statement when  `sliceToWorkers` isn't empty (so some slices / worker groups have already been added) and we intercept the first pod in a different worker group.

File: applications/ray/kuberay-tpu-webhook/main.go (hunk 5 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go:333]: By "created" we just mean that the slice (i.e. a replica of a TPU worker group) has been added to `sliceToWorkers`, in the case where `NumOfHosts` workers have been added to `sliceToWorkers` for the other replicas of the worker group, `numSlices`  (I'll change it to `numReplicas`) indicates the next `replicaIndex` to assign.

File: applications/ray/kuberay-tpu-webhook/main.go (hunk 3 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go]: Done.

File: applications/ray/kuberay-tpu-webhook/main.go (hunk 4 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go:342]: Done.

File: applications/ray/kuberay-tpu-webhook/main.go (hunk 1 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/main.go:80]: I can move the log statements somewhere else, but we should always have `clusterName`, `namespace`, and `groupName` (if not it'd return an error before `getNumTPUHostsFromTopology` so it felt safe to pass them to the function.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go [hunk 7]
```diff
@@ -0,0 +1,595 @@
...
+
+func setupTest(t *testing.T) {
+	namespaceStr = "default"
+	instanceName = "raycluster-sample"
+	headNodeIP = "1.2.3.4"
+	groupNameStr = "workergroup"
+	headlessServiceSuffix = "headless-worker-svc"
+
+	// 1 CPU head pod + 1 worker - doesn't request TPUs
+	testCPUPods = []runtime.Object{
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/webhook_main_test.go:48]: Is there a way we could define constant yamls for something like "cpu worker" and "tpu worker", and then re-use these templates in subsequent tests? 

One option is to add a "testdata" directory where you could define these as yaml strings, which you could then parse here to reduce the repetition in test code.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go [hunk 8]
```diff
@@ -0,0 +1,595 @@
...
+		},
+	}
+}
+
+func Test_GetReplicaIndex(t *testing.T) {
+	setupTest(t)
+
+	tests := map[string]struct {
+		sliceToWorkers     map[slice][]worker
+		numOfHosts         int32
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/webhook_main_test.go:460]: Maybe call these "expectedNumOfHosts" and "expectedNumReplicas" to be more clear.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go [hunk 9]
```diff
@@ -0,0 +1,595 @@
...
+		numReplicas        int
+		additionalGroupStr string
+		numOfHosts2        int32
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/webhook_main_test.go:463]: expectedAdditionalNumOfHosts? to make the association with `additionalGroup` more explicit.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go [hunk 10]
```diff
@@ -0,0 +1,595 @@
...
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func Test_GetReplicaIndex(t *testing.T) {
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/webhook_main_test.go:455]: One interesting way to test this function is to override the contents of `sliceToWorkers` and seeing how this function behaves in different circumstances.

We can do this in a follow up PR.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go [hunk 11]
```diff
@@ -0,0 +1,595 @@
...
+			// for each respective worker group
+			numOfHosts:         4,
+			numReplicas:        4,
+			additionalGroupStr: "another-worker-group",
+			numOfHosts2:        2,
+			numReplicas2:       3,
+		},
+	}
+
+	for name, tc := range tests {
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/webhook_main_test.go:497]: Describe what the test is trying to validate.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go [hunk 12]
```diff
@@ -0,0 +1,595 @@
...
+func Test_GetNextWorkerID(t *testing.T) {
+	setupTest(t)
+
+	tests := map[string]struct {
+		numOfHosts  int32
+		numReplicas int
+		deletePodID int
+	}{
+		"single-host, single-slice worker group": {
+			// single-slice, replicaIndex should always be 0
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/webhook_main_test.go:546]: What should the worker ids be here?

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go [hunk 13]
```diff
@@ -0,0 +1,595 @@
...
+			// multi-slice, replicaIndex should always be 0-numReplicas for 0-numOfHosts pods
+			numOfHosts:  4,
+			numReplicas: 4,
+			deletePodID: -1,
+		},
+		"deleted pod from multi-host group": {
+			// pod should be reassigned the ID of the deleted pod
+			numOfHosts:  4,
+			numReplicas: 4,
+			deletePodID: 2,
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/webhook_main_test.go:573]: What if you delete multiple pods?

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go [hunk 14]
```diff
@@ -0,0 +1,595 @@
...
+		},
+	}
+
+	for name, tc := range tests {
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/webhook_main_test.go:577]: Same thing here, describe what this test validates.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go [hunk 15]
```diff
@@ -0,0 +1,595 @@
...
+	testCPUPods                   []runtime.Object
+	testTPUPods                   []runtime.Object
+	testRayClusterAdmissionReview *admissionv1.AdmissionReview
+	testRayClusterNoTPUs          *rayv1.RayCluster
+	testRayClusterSingleHostTPU   *rayv1.RayCluster
+	testRayClusterMultiHostTPU    *rayv1.RayCluster
+	testServices                  []runtime.Object
+	workerSelector                labels.Selector
+	headNodeIP                    string
+	// sliceToWorkers				  map[slice][]worker
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/webhook_main_test.go:35]: Is this intended to use the same copy of `sliceToWorkers` in the webhook?

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go [hunk 16]
```diff
@@ -0,0 +1,595 @@
...
+			numReplicas: 1,
+			deletePodID: -1,
+		},
+		"multi-host, multi-slice worker group": {
+			// multi-slice, replicaIndex should always be 0-numReplicas for 0-numOfHosts pods
+			numOfHosts:  4,
+			numReplicas: 4,
+			deletePodID: -1,
+		},
+		"deleted pod from multi-host group": {
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/webhook_main_test.go:569]: Do we need a test for multiple worker groups here?

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go [hunk 17]
```diff
@@ -0,0 +1,595 @@
...
+			// single-slice, replicaIndex should always be 0
+			numOfHosts:  4,
+			numReplicas: 1,
+		},
+		"multi-host, multi-slice worker group": {
+			// multi-slice, replicaIndex should always be 0-numReplicas for 0-numOfHosts pods
+			numOfHosts:  4,
+			numReplicas: 4,
+		},
+		"multiple worker groups": {
```
Comment (richardsliu) [inline applications/ray/kuberay-tpu-webhook/webhook_main_test.go:486]: Do we need a test for deleting pods here?

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go (hunk 7 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/webhook_main_test.go]: I think that's a good idea, for now I changed `testCPUPods` and `testTPUPods` to a cpu and tpu worker template whose values we can change within the test. In the next PR with the rest of the tests I think it'd be good to add a testdata directory (since these objects aren't being used yet in this PR).

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go (hunk 8 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/webhook_main_test.go]: These aren't the expected values, these define the test values for `NumOfHosts` and `Replicas` for a hypothetical Ray worker group. The expected values will then be between `0-(NumOfHosts-1)` and `0-(Replicas-1)`.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go (hunk 9 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/webhook_main_test.go]: Done, changed to `additionalNumOfHosts`.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go (hunk 11 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/webhook_main_test.go:188]: Done.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go (hunk 12 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/webhook_main_test.go]: Done, changed the comment descriptions to define what the worker IDs should be for each test.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go (hunk 13 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/webhook_main_test.go]: Done, added a test case for this.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go (hunk 14 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/webhook_main_test.go:359]: Done.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go (hunk 15 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/webhook_main_test.go]: Yeah it uses the same `sliceToWorkers` since this is what's added to internally in `getNextWorkerID`. I'll add in some logic to save a deep copy of `sliceToWorkers` and return it to its state after the test. I don't think this should cause any issues unless you run the tests while the webhook is intercepting pods with the same cluster name, namespace, and group name.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go (hunk 16 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/webhook_main_test.go]: Done, added a test case for this to validate that re-created pods will be assigned the correct ID in different worker groups.

File: applications/ray/kuberay-tpu-webhook/webhook_main_test.go (hunk 17 above)
Comment (ryanaoleary) [reply applications/ray/kuberay-tpu-webhook/webhook_main_test.go:160]: Added a test case to validate that a re-created pod will be assigned to the lowest index replica with created Pods < NumOfHosts.

**Review by richardsliu — APPROVED** [review]
